### PR TITLE
breaking: turn `redirect` and `error` into commands. Export helpers for identifying them when caught

### LINKED ISSUE #6009: HttpError and Redirect import errors
### Describe the bug

I'm building my own adapter with typescript, and I use `tsc -b` to build it.
Since the new routing change, I get a type error when building (see reproduction).

Also: `HttpError` and `Redirect` classes are not available to import, which they probably should be since we indirectly throw them and might want to handle them differently if thrown around with other errors.

See also Rich's comment here: https://github.com/sveltejs/kit/issues/5952#issuecomment-1218844057

Also also, not directly related but still weird typing issue: I had to add `@types/estree` to dependencies to appease svelte, seems related to: https://github.com/sveltejs/svelte/issues/3397

### Reproduction

https://stackblitz.com/edit/node-sfis81?file=index.ts

### Logs

_No response_

### System Info

```shell
System:
    OS: Windows 10 10.0.19044
    CPU: (8) x64 Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz    
    Memory: 1.41 GB / 15.75 GB
  Binaries:
    Node: 16.16.0 - ~\AppData\Local\nvs\default\node.EXE     
    npm: 8.11.0 - ~\AppData\Local\nvs\default\npm.CMD        
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (104.0.1293.54)
    Internet Explorer: 11.0.19041.1566
  npmPackages:
    @sveltejs/kit: 1.0.0-next.420 => 1.0.0-next.420 
    svelte: ^3.49.0 => 3.49.0 
    vite: ^3.0.8 => 3.0.8
```

### Severity

serious, but I can work around it

### Additional Information

_No response_

### LINKED ISSUE #11144: Turn `error` and `redirect` into commands, rather than functions returning a value we have to throw
### Describe the problem

Throwing `error` and `redirect` has always been a bit weird, and people have complained about it before, but it _is_ necessary -- both of them are meant to stop the normal control flow and start something new, which is kinda what `throw` does. Remembering that you have to throw them, though, is strange, and it doesn't feel like a great developer API. Thankfully, TypeScript gives us what we need to have the best of both worlds:

```ts
function redirect(): never {}
```

A function with a return type of `never`, well, _never_ returns. TypeScript treats it like a thrown error. In the following code:

```ts
const foo = 'bar';
redirect('/hello-world');
console.log('hello, world');
```

TypeScript would grey out the last line, knowing that it could never be reached. This is ultra-simple to implement, and would basically just mean changing the implementation of the current `redirect` and `error` functions like so:

```ts
// old
function redirect() {
  return new Redirect();
}

// new
function redirect(): never {
  throw new Redirect();
}
```

### Describe the proposed solution

Both combined above.

### Alternatives considered

Leaving it as-is. The main downside here is that users can't do something like:

```ts
const e = error(401);
throw e;
```

but the only reason they can do this today is because we didn't think of the above before this feature was released -- the actual redirect and error classes are meant to be implementation details not exposed to the user.

### Importance

nice to have

### Additional Information

_No response_

### LINKED ISSUE #11186: Typescript eslint error "no-throw-literal" on error or redirect functions
### Describe the bug

Hello :wave:,

I'm new with svelte and I have a lint error that I fail to correct. When using `throw error...` or `throw redirect...` eslint report an error `@typescript-eslint/no-throw-literal`

Thanks in advance.

### Reproduction

You can reproduce the error on [this repository](https://github.com/Its-Alex/curriculum/), by removing [this comment](https://github.com/Its-Alex/curriculum/blob/master/src/routes/+error.svelte#L6)

I use pnpm and node, you can find versions in [.tool-versions](https://github.com/Its-Alex/curriculum/blob/master/.tool-versions)

### Logs

_No response_

### System Info

```Shell
$ pnpx envinfo --system --binaries --browsers --npmPackages "{svelte,@sveltejs/*,vite}"
Packages: +1
+
Progress: resolved 1, reused 0, downloaded 1, added 1, done

  System:
    OS: Linux 6.6 Arch Linux
    CPU: (24) x64 AMD Ryzen 9 5900X 12-Core Processor
    Memory: 22.44 GB / 31.27 GB
    Container: Yes
    Shell: 5.9 - /usr/bin/zsh
  Binaries:
    Node: 20.10.0 - ~/.local/share/rtx/installs/node/20.10.0/bin/node
    npm: 10.2.3 - ~/.local/share/rtx/installs/node/20.10.0/bin/npm
    pnpm: 8.11.0 - ~/.local/share/rtx/installs/pnpm/8.11.0/bin/pnpm
  Browsers:
    Chromium: 119.0.6045.199
  npmPackages:
    @sveltejs/adapter-node: 1.3.1 => 1.3.1
    @sveltejs/adapter-vercel: 3.1.0 => 3.1.0
    @sveltejs/kit: 1.27.6 => 1.27.6
    svelte: 4.2.8 => 4.2.8
    vite: 4.5.0 => 4.5.0
```

### Severity

annoyance

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by tcc-sejohnson**
Closes #11144
Closes #6009
Closes #11186

This makes `redirect` and `error` into functions that return `never`, meaning we can throw errors from inside of them without making our users do it.

This also adds `isHttpError` and `isRedirect` helpers to identify these commands if you've caught them -- for example:

```ts
export function withLogging(loadFn): loadFn {
  return () => {
    try {
      return await loadFn();
    } catch (e) {
      if (!isRedirect(e)) {
        logError(e);
      }
      throw e;
    }
  }
}
```

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

File: .changeset/clean-cars-kiss.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@sveltejs/kit': major
+---
+
+breaking: Turn `error` and `redirect` into commands
```
Comment (benmccann) [inline .changeset/clean-cars-kiss.md:5]: we've been starting these with a lowercase letter
```suggestion
breaking: turn `error` and `redirect` into commands
```

File: packages/kit/src/exports/index.js [hunk 2]
```diff
@@ -31,21 +29,46 @@ export function error(status, body) {
...
 	if ((!BROWSER || DEV) && (isNaN(status) || status < 300 || status > 308)) {
 		throw new Error('Invalid status code');
 	}
 
-	return new Redirect(status, location.toString());
+	throw new Redirect(status, location.toString());
+}
+
+/**
+ * Checks whether this is a redirect thrown by {@link redirect}.
```
Comment (eltigerchino) [inline packages/kit/src/exports/index.js:66]: see https://github.com/sveltejs/kit/pull/11165/files#r1413875219

File: packages/kit/src/exports/index.js [hunk 3]
```diff
@@ -31,21 +29,46 @@ export function error(status, body) {
 		throw new Error(`HTTP error status codes must be between 400 and 599 — ${status} is invalid`);
 	}
 
-	return new HttpError(status, body);
+	throw new HttpError(status, body);
 }
 
 /**
- * Create a `Redirect` object. If thrown during request handling, SvelteKit will return a redirect response.
+ * Checks whether this is an error thrown by {@link error}.
```
[the remark this answers is not in the document]
Comment (tcc-sejohnson) [reply packages/kit/src/exports/index.js:36]: So weird that this is causing a circular error... the definition for `error` doesn't refer to this function. This has got to be a bug with whatever we're using to build our docs.

[the remark this answers is not in the document]
Comment (eltigerchino) [reply packages/kit/src/exports/index.js:36]: Yup, you’re right. There’s an error when it comes to parsing the jsdoc comment for it

[the remark this answers is not in the document]
Comment (eltigerchino) [reply packages/kit/src/exports/index.js:47]: Added a PR to fix this in master which can then be pulled into this PR https://github.com/sveltejs/kit/pull/11193

**Comment by dummdidumm** [comment]
Added a migration script for this in #11199

**Comment by benmccann** [comment]
Should we use the `@throws` JSDoc tag on these declarations? I don't know how much it might help or not, but it could make it slightly more visible that it throws an error

**Comment by tcc-sejohnson** [comment]
@benmccann Oh, that might not be a bad idea. Someone can do it or I'll try to get to it today

**Comment by tcc-sejohnson** [comment]
@eltigerchino 

I think I broke the JSDoc for `error`, not sure how -- the written documentation is no longer showing when I hover over the function. I think you'd fixed this before but 🤷🏻 I can't figure it out.

**Comment by eltigerchino** [comment]
> @eltigerchino 
> 
> 
> 
> I think I broke the JSDoc for `error`, not sure how -- the written documentation is no longer showing when I hover over the function. I think you'd fixed this before but 🤷🏻 I can't figure it out.

Oh, it's a typescript issue where the descriptions are ignored when you have the overload tag. You just need to repeat the definitions that are beneath the overload tag at the top to get it working again.

e.g.,
```
/**
* Hello world!
* @params {string} something - Pass me anything
* @overload
* @params {string} something
*/
```

We have to do this for each overload definition.

**Comment by tcc-sejohnson** [comment]
> > @eltigerchino
> > I think I broke the JSDoc for `error`, not sure how -- the written documentation is no longer showing when I hover over the function. I think you'd fixed this before but 🤷🏻 I can't figure it out.
> 
> Oh, it's a typescript issue where all the descriptions are ignored when you have the overload tag. You just need to repeat the definitions with the descriptions before the overload tag to get it working again.
> 
> e.g.,
> 
> ```
> /**
> * Hello world!
> * @params {string} something - Pass me anything
> * @overload
> * @params {string} something
> */
> ```
> 
> We have to do this for each overload definition.

Yeah... I just tried this and it's not working. Adding the @throw tags might have broken it once and for all 😅

**Comment by eltigerchino** [comment]
It's strange because the `@throws` error type gets stripped out by dts-buddy when generating the types? also, it's not rendered correctly properly in VSCode although it's supposedly fixed in TypeScript https://github.com/microsoft/TypeScript/issues/49323

error type is not rendered correctly

Curiously, the `@throws` JSDoc comments don't need to be duplicated, but the others do. Maybe we need to fix something in dts-buddy

**Comment by eltigerchino** [comment]
Another alternative is to wrap the type in backticks until it gets fixed.
```diff
- @throws {HttpError} This error instructs SvelteKit to initiate HTTP error handling.
+ @throws `HttpError` This error instructs SvelteKit to initiate HTTP error handling.
```
CleanShot 2023-12-06 at 11  28 13@2x

**Comment by benmccann** [comment]
It looks like this PR got corrupted with other changes included on it. It's deleting a few changesets and has other changes that I don't think should be included

**Comment by Rich-Harris** [comment]
I fixed the `@throws` thing. The duplicate documentation is annoying, but is an issue we [already have on the site](https://kit.svelte.dev/docs/modules#sveltejs-kit-error) so I don't think it needs to block this PR

**Review by Rich-Harris — APPROVED** [review]

**Pull request merged**